### PR TITLE
refactor: Revise script structure

### PR DESCRIPTION
**Pull request by xxuejie**
* [x] rfc changes: https://github.com/nervosnetwork/rfcs/pull/91
* [x] ckb-system-scripts: https://github.com/nervosnetwork/ckb-system-scripts/pull/2
* [x] mruby-contracts: https://github.com/nervosnetwork/mruby-contracts/pull/7
* [x] ruby sdk: https://github.com/nervosnetwork/ckb-demo-ruby/pull/76
* [x] ruby scripts: https://github.com/nervosnetwork/ckb-ruby-scripts/pull/9
* [ ] swift sdk
* [ ] java sdk
* [ ] js sdk

---

This is a **BREAKING CHANGE**.

### CODE REVIEW

File: core/src/script.rs [hunk 1]
```diff
@@ -13,54 +13,11 @@ use std::mem;
...
-        match self.0 {
-            Some(ref v) => write!(f, "Some({})", v),
-            None => write!(f, "None"),
-        }
-    }
+    // The reference hash here can be used to refer to binary in any of the
+    // following locations:
+    // 1. Data part of a dep cell in current transaction
+    // 2. An embed item in current transaction
+    pub reference: H256,
```
Comment (Mine77) [inline core/src/script.rs]: Could you please elaborate what exact hash should be used here to refer a dep cell or an embed item in tx?

Is it the hash of the `data` part of a cell?

Comment (xxuejie) [reply core/src/script.rs]: Let me update the comment to be more clear

File: core/src/script.rs [hunk 2]
```diff
@@ -13,54 +13,13 @@ use std::mem;
...
-        match self.0 {
-            Some(ref v) => write!(f, "Some({})", v),
-            None => write!(f, "None"),
-        }
-    }
+    // The reference hash here can be used to refer to binary in any of the
+    // following locations:
+    // 1. Data part of a dep cell in current transaction, reference hash
+    // must be the hash of the cell data so as to reference this cell
+    // 2. An embed item in current transaction, reference hash must be the
```
Comment (u2) [inline core/src/script.rs]: So the embed item is a one-time data?

Comment (xxuejie) [reply core/src/script.rs]: Yep, it's like the original binary but at a different place with more potential use cases

File: verification/src/transaction_verifier.rs [hunk 3]
```diff
@@ -70,18 +70,8 @@ impl<'a> InputVerifier<'a> {
...
-                if let Some(ref input_cell) = cs.get_live() {
-                    // TODO: remove this once VM mmap is in place so we can
-                    // do P2SH within the VM.
-                    if input_cell.lock != input.unlock.type_hash() {
-                        return Err(TransactionError::InvalidScript);
-                    }
-                }
-            } else if cs.is_dead() {
+        for cs in &self.resolved_transaction.input_cells {
+            if cs.is_dead() {
```
Comment (u2) [inline verification/src/transaction_verifier.rs]: :+1:

**Comment by u2** [comment]
Add some description for this PR. The linked issue is private.

File: core/src/script.rs (hunk 1 above)
Comment (doitian) [reply core/src/script.rs]: Is `binary_hash` a better name?

File: script/src/syscalls/mod.rs [hunk 4]
```diff
@@ -29,6 +29,7 @@ enum CellField {
     Capacity = 0,
     Data = 1,
     DataHash = 2,
+    Lock = 6,
```
Comment (jjyr) [inline script/src/syscalls/mod.rs]: Move this item after `TypeHash = 5` for consistency.

Comment (xxuejie) [reply script/src/syscalls/mod.rs]: There are 2 ways to understand consistency here: first, we can sort them by the value of the enums, of course. But there's also another way which we can group the values based on the meeting, so we can do Data, DataHash, Lock, LockHash, this way it's easier to see what options are available here. Thoughts?

**Comment by doitian** [comment]
@xxuejie this PR conflicts with the #350

**Comment by xxuejie** [comment]
@doitian Conflict is resolved now

**Comment by ashchan** [comment]
SDKs haven't caught up yet, but I guess we can plan this for `v0.8.0` which is RC frozen today (meaning develop during the RC period). How about that? @doitian @xxuejie

**Comment by Mine77** [comment]
are you considering updating the [how-to-write-contract doc](https://github.com/nervosnetwork/ckb-demo-ruby/blob/develop/docs/how-to-write-contracts.md) as well?

**Comment by doitian** [comment]
Integration test stuck, cc @quake

**Comment by xxuejie** [comment]
bors try

**Pull request merged**